File: android/app/src/main/java/com/danusha/tgwsproxy/NativeBridge.kt
package com.danusha.tgwsproxy

import org.json.JSONObject

object NativeBridge {
    init {
        System.loadLibrary("tg_ws_proxy_android")
    }

    @JvmStatic
    private external fun nativeStart(configJson: String): String

    @JvmStatic
    private external fun nativeStop(): String

    @JvmStatic
    private external fun nativeStatus(): String

    fun start(configJson: String): NativeResponse = NativeResponse.parse(nativeStart(configJson))

    fun stop(): NativeResponse = NativeResponse.parse(nativeStop())

    fun status(): ProxyStatus = ProxyStatus.parse(nativeStatus())
}

data class NativeResponse(
    val ok: Boolean,
    val error: String?,
) {
    companion object {
        fun parse(json: String): NativeResponse {
            val value = JSONObject(json)
            return NativeResponse(
                ok = value.optBoolean("ok", false),
                error = value.optString("error").takeIf { it.isNotBlank() && it != "null" },
            )
        }
    }
}

data class ProxyStatus(
    val state: String,
    val error: String?,
    val telegramUrl: String?,
    val totalConnections: Long,
    val activeConnections: Long,
    val bytesUp: Long,
    val bytesDown: Long,
) {
    val isActive: Boolean
        get() = state == "starting" || state == "running" || state == "stopping"

    companion object {
        fun parse(json: String): ProxyStatus {
            val value = JSONObject(json)
            return ProxyStatus(
                state = value.optString("state", "failed"),
                error = value.optString("error").takeIf { it.isNotBlank() && it != "null" },
                telegramUrl = value.optString("telegramUrl")
                    .takeIf { it.isNotBlank() && it != "null" },
                totalConnections = value.optLong("totalConnections"),
                activeConnections = value.optLong("activeConnections"),
                bytesUp = value.optLong("bytesUp"),
                bytesDown = value.optLong("bytesDown"),
            )
        }
    }
}

File: ios/TgWsProxy/Proxy/NativeBridge.swift
import Foundation

/// Response returned by `tgws_start` and `tgws_stop`.
struct NativeResponse: Decodable {
    let ok: Bool
    let error: String?
}

/// Runtime state and traffic counters reported by the Rust core.
struct ProxyStatus: Equatable {
    var state: String = "stopped"
    var error: String?
    var telegramUrl: String?
    var startedAtEpochSeconds: UInt64?
    var totalConnections: UInt64 = 0
    var activeConnections: UInt64 = 0
    var websocketConnections: UInt64 = 0
    var tcpFallbackConnections: UInt64 = 0
    var cloudflareConnections: UInt64 = 0
    var badConnections: UInt64 = 0
    var bytesUp: UInt64 = 0
    var bytesDown: UInt64 = 0

    /// True while the core owns the listener socket, including the transitions.
    var isActive: Bool {
        state == "starting" || state == "running" || state == "stopping"
    }

    var isRunning: Bool { state == "running" }
}

extension ProxyStatus: Decodable {
    private enum CodingKeys: String, CodingKey {
        case state, error, telegramUrl, startedAtEpochSeconds
        case totalConnections, activeConnections, websocketConnections
        case tcpFallbackConnections, cloudflareConnections, badConnections
        case bytesUp, bytesDown
    }

    /// Decodes defensively: a core that fails while serialising its status
    /// sends `state` and `error` only.
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        func counter(_ key: CodingKeys) throws -> UInt64 {
            try container.decodeIfPresent(UInt64.self, forKey: key) ?? 0
        }
        state = try container.decodeIfPresent(String.self, forKey: .state) ?? "failed"
        error = try container.decodeIfPresent(String.self, forKey: .error)
        telegramUrl = try container.decodeIfPresent(String.self, forKey: .telegramUrl)
        startedAtEpochSeconds = try container.decodeIfPresent(
            UInt64.self,
            forKey: .startedAtEpochSeconds
        )
        totalConnections = try counter(.totalConnections)
        activeConnections = try counter(.activeConnections)
        websocketConnections = try counter(.websocketConnections)
        tcpFallbackConnections = try counter(.tcpFallbackConnections)
        cloudflareConnections = try counter(.cloudflareConnections)
        badConnections = try counter(.badConnections)
        bytesUp = try counter(.bytesUp)
        bytesDown = try counter(.bytesDown)
    }
}

/// Thin wrapper over the C ABI of `crates/ios-bridge`.
enum NativeBridge {
    static func start(configurationJSON: String) -> NativeResponse {
        let raw = configurationJSON.withCString { pointer in
            takeString(tgws_start(pointer))
        }
        return decode(raw)
    }

    static func stop() -> NativeResponse {
        decode(takeString(tgws_stop()))
    }

    static func status() -> ProxyStatus {
        let raw = takeString(tgws_status())
        guard
            let data = raw.data(using: .utf8),
            let status = try? JSONDecoder().decode(ProxyStatus.self, from: data)
        else {
            var status = ProxyStatus()
            status.state = "failed"
            status.error = String(localized: "The core returned an unreadable status.")
            return status
        }
        return status
    }

    /// Copies a string returned by the core and releases the native allocation.
    private static func takeString(_ pointer: UnsafeMutablePointer<CChar>?) -> String {
        guard let pointer else { return "" }
        defer { tgws_free_string(pointer) }
        return String(cString: pointer)
    }

    private static func decode(_ raw: String) -> NativeResponse {
        guard
            let data = raw.data(using: .utf8),
            let response = try? JSONDecoder().decode(NativeResponse.self, from: data)
        else {
            return NativeResponse(
                ok: false,
                error: String(localized: "The core returned an unreadable response.")
            )
        }
        return response
    }
}
